// !LANGUAGE: +SafeExternalEnums
// FIR_IDENTICAL
external interface Testable {
    fun test(): Int
}
external open class CustomClass

external enum class Test1 { A, B, C }
external enum class Test2: Enum<Test2> { A, B, C }
external enum class Test3: <!CLASS_CANNOT_BE_EXTENDED_DIRECTLY!>ExternalEnum<!> { A, B, C }
external enum class Test4: <!CLASS_IN_SUPERTYPE_FOR_ENUM, FINAL_SUPERTYPE!>Test2<!> { A, B, C }
external enum class Test5: Enum<Test5>, Testable { <!ABSTRACT_MEMBER_NOT_IMPLEMENTED!>A<!>, <!ABSTRACT_MEMBER_NOT_IMPLEMENTED!>B<!>, <!ABSTRACT_MEMBER_NOT_IMPLEMENTED!>C<!> }
external enum class Test6: Testable, Enum<Test6>  { <!ABSTRACT_MEMBER_NOT_IMPLEMENTED!>A<!>, <!ABSTRACT_MEMBER_NOT_IMPLEMENTED!>B<!>, <!ABSTRACT_MEMBER_NOT_IMPLEMENTED!>C<!> }
external enum class Test7: Enum<Test7>, <!CLASS_IN_SUPERTYPE_FOR_ENUM, MANY_CLASSES_IN_SUPERTYPE_LIST!>CustomClass<!>  { A, B, C }
external enum class Test8: <!CLASS_IN_SUPERTYPE_FOR_ENUM!>CustomClass<!>  { A, B, C }
external enum class Test9: <!CLASS_IN_SUPERTYPE_FOR_ENUM!>CustomClass<!>, Testable  { <!ABSTRACT_MEMBER_NOT_IMPLEMENTED!>A<!>, <!ABSTRACT_MEMBER_NOT_IMPLEMENTED!>B<!>, <!ABSTRACT_MEMBER_NOT_IMPLEMENTED!>C<!> }
external enum class Test10: <!CLASS_IN_SUPERTYPE_FOR_ENUM!>CustomClass<!>, <!CLASS_CANNOT_BE_EXTENDED_DIRECTLY!>ExternalEnum<!>  { A, B, C }
external enum class Test11: <!CLASS_CANNOT_BE_EXTENDED_DIRECTLY!>ExternalEnum<!>, Enum<Test11> { A, B, C }

enum class Test12 { A, B, C }
enum class Test13: <!CLASS_CANNOT_BE_EXTENDED_DIRECTLY, SUPERTYPE_NOT_INITIALIZED!>Enum<Test13><!> { A, B, C }
enum class Test14: <!CLASS_CANNOT_BE_EXTENDED_DIRECTLY!>ExternalEnum<!> { A, B, C }
enum class Test15: <!CLASS_IN_SUPERTYPE_FOR_ENUM, FINAL_SUPERTYPE, SUPERTYPE_NOT_INITIALIZED!>Test2<!> { A, B, C }
enum class Test16: <!CLASS_CANNOT_BE_EXTENDED_DIRECTLY, SUPERTYPE_NOT_INITIALIZED!>Enum<Test16><!>, Testable { <!ABSTRACT_MEMBER_NOT_IMPLEMENTED!>A<!>, <!ABSTRACT_MEMBER_NOT_IMPLEMENTED!>B<!>, <!ABSTRACT_MEMBER_NOT_IMPLEMENTED!>C<!> }
enum class Test17: Testable, <!CLASS_CANNOT_BE_EXTENDED_DIRECTLY, SUPERTYPE_NOT_INITIALIZED!>Enum<Test17><!>  { <!ABSTRACT_MEMBER_NOT_IMPLEMENTED!>A<!>, <!ABSTRACT_MEMBER_NOT_IMPLEMENTED!>B<!>, <!ABSTRACT_MEMBER_NOT_IMPLEMENTED!>C<!> }
enum class Test18: <!CLASS_CANNOT_BE_EXTENDED_DIRECTLY, SUPERTYPE_NOT_INITIALIZED!>Enum<Test18><!>, <!CLASS_IN_SUPERTYPE_FOR_ENUM, MANY_CLASSES_IN_SUPERTYPE_LIST, SUPERTYPE_NOT_INITIALIZED!>CustomClass<!>  { A, B, C }
enum class Test19: <!CLASS_IN_SUPERTYPE_FOR_ENUM, SUPERTYPE_NOT_INITIALIZED!>CustomClass<!>  { A, B, C }
enum class Test20: <!CLASS_IN_SUPERTYPE_FOR_ENUM, SUPERTYPE_NOT_INITIALIZED!>CustomClass<!>, Testable  { <!ABSTRACT_MEMBER_NOT_IMPLEMENTED!>A<!>, <!ABSTRACT_MEMBER_NOT_IMPLEMENTED!>B<!>, <!ABSTRACT_MEMBER_NOT_IMPLEMENTED!>C<!> }
enum class Test21: <!CLASS_IN_SUPERTYPE_FOR_ENUM, SUPERTYPE_NOT_INITIALIZED!>CustomClass<!>, <!CLASS_CANNOT_BE_EXTENDED_DIRECTLY!>ExternalEnum<!>  { A, B, C }
enum class Test22: <!CLASS_CANNOT_BE_EXTENDED_DIRECTLY!>ExternalEnum<!>, <!CLASS_CANNOT_BE_EXTENDED_DIRECTLY, SUPERTYPE_NOT_INITIALIZED!>Enum<Test22><!> { A, B, C }
